src/txn/txn.c

```diff
@@ -1465,32 +1465,99 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
     return (ret);
 }
 
+/*
+ * __txn_mod_sortable_key --
+ *     Given an operation return a boolean indicating if it has a sortable key.
+ */
+static inline bool
+__txn_mod_sortable_key(WT_SESSION_IMPL *session, WT_TXN_OP *opt)
+{
+    switch (opt->type) {
+    case (WT_TXN_OP_NONE):
+    case (WT_TXN_OP_REF_DELETE):
+    case (WT_TXN_OP_TRUNCATE_COL):
+    case (WT_TXN_OP_TRUNCATE_ROW):
+        return (false);
+    case (WT_TXN_OP_BASIC_COL):
+    case (WT_TXN_OP_BASIC_ROW):
+    case (WT_TXN_OP_INMEM_COL):
+    case (WT_TXN_OP_INMEM_ROW):
+        return (true);
+    }
+    WT_ASSERT_ALWAYS(session, false, "Unhandled op type encountered.");
+    return (false);
+}
+
 /*
  * __txn_mod_compare --
- *     Qsort comparison routine for transaction modify list.
+ *     Qsort comparison routine for transaction modify list. Takes a session as a context argument.
+ *     This allows for the use of comparators.
  */
 static int WT_CDECL
-__txn_mod_compare(const void *a, const void *b)
+__txn_mod_compare(const void *a, const void *b, void *context)
 {
+    WT_SESSION_IMPL *session;
     WT_TXN_OP *aopt, *bopt;
+    int cmp;
+    bool a_has_sortable_key;
+    bool b_has_sortable_key;
 
     aopt = (WT_TXN_OP *)a;
     bopt = (WT_TXN_OP *)b;
+    session = (WT_SESSION_IMPL *)context;
 
-    /* If the files are different, order by ID. */
-    if (aopt->btree->id != bopt->btree->id)
-        return (aopt->btree->id < bopt->btree->id);
+    /*
+     * We want to sort on two things:
+     *  - B-tree ID
+     *  - Key
+     * However, there are a number of modification types that don't have a key to be sorted on. This
+     * requires us to add a stage between sorting on B-tree ID and key. At this intermediate stage,
+     * we sort on whether the modifications have a key.
+     *
+     * We need to uphold the contract that all modifications on the same key are contiguous in the
+     * final modification array. Technically they could be separated by non key modifications,
+     * but for simplicity's sake we sort them apart.
+     *
+     * Qsort comparators are expected to return -1 if the first argument is smaller than the second,
+     * 1 if the second argument is smaller than the first, and 0 if both arguments are equal.
+     */
+
+    /* Order by b-tree ID. */
+    if (aopt->btree->id < bopt->btree->id)
+        return (-1);
+    if (bopt->btree->id < aopt->btree->id)
+        return (1);
 
     /*
-     * If the files are the same, order by the key. Row-store collators require WT_SESSION pointers,
-     * and we don't have one. Compare the keys if there's no collator, otherwise return equality.
-     * Column-store is always easy.
+     * Order by whether the given operation has a key. We don't want to call key compare incorrectly
+     * especially given that u is a union which would create undefined behavior.
+     */
+    a_has_sortable_key = __txn_mod_sortable_key(session, aopt);
```

Comment:
To confirm my understanding, from the logic, all the ops with keys are sorted first and followed by ops with no keys for the same btree. As we are sorting these ops to mark all the modifications for the same key to set their timestamps before we release the page. The non-op operations don't search for a page, it should be fine even if they are set at the end.

Reply:
Yes, that's correct.

Reply:
Thanks.